File: src/data/repositories/MatchRepository.ts
import { Match } from "@domain/models/match";

export interface MatchRepository {
  list(): Promise<Match[]>;
  get(id: string): Promise<Match | undefined>;
  upsert(match: Match): Promise<void>;
  remove(id: string): Promise<void>;
}

export class InMemoryMatchRepository implements MatchRepository {
  private items: Match[] = [];

  constructor(seed?: Match[]) {
    if (seed?.length) {
      this.items = seed;
    } else {
      const now = new Date();
      this.items = [
        {
          id: "1",
          dateISO: new Date(now.getFullYear(), now.getMonth(), now.getDate(), 20, 0, 0).toISOString(),
          location: "서울 ○○풋살장 A코트",
          durationMin: 90,
          goals: 2,
          assists: 1,
          rating: 7,
          memo: "오랜만에 몸 풀림"
        }
      ];
    }
  }

  async list(): Promise<Match[]> {
    return [...this.items].sort((a, b) => (a.dateISO < b.dateISO ? 1 : -1));
  }

  async get(id: string): Promise<Match | undefined> {
    return this.items.find(m => m.id === id);
  }

  async upsert(match: Match): Promise<void> {
    const idx = this.items.findIndex(m => m.id === match.id);
    if (idx >= 0) {
      this.items[idx] = match;
    } else {
      this.items.push(match);
    }
  }

  async remove(id: string): Promise<void> {
    this.items = this.items.filter(m => m.id !== id);
  }
}

export const matchRepository = new InMemoryMatchRepository();

File: BallLog_v1/BallLog_v1/Data/MatchRepository.swift
import Foundation

protocol MatchRepository {
    func list() -> [Match]
    func get(_ id: UUID) -> Match?
    func upsert(_ match: Match)
    func remove(_ id: UUID)
}

final class InMemoryMatchRepository: MatchRepository {
    private var items: [Match] = []
    
    init(seed: [Match]? = nil) {
        if let seed, !seed.isEmpty {
            self.items = seed
        } else {
            let now = Date()
            self.items = [
                Match(
                    id: UUID(),
                    date: Calendar.current.date(bySettingHour: 20, minute: 0, second: 0, of: now) ?? now,
                    location: "서울 ○○풋살장 A코트",
                    durationMin: 90,
                    goals: 2,
                    assists: 1,
                    rating: 7,
                    memo: "오랜만에 몸 풀림"
                )
            ]
        }
    }
    
    func list() -> [Match] {
        items.sorted { $0.date > $1.date }
    }
    
    func get(_ id: UUID) -> Match? {
        items.first(where: { $0.id == id })
    }
    
    func upsert(_ match: Match) {
        if let idx = items.firstIndex(where: { $0.id == match.id }) {
            items[idx] = match
        } else {
            items.append(match)
        }
    }
    
    func remove(_ id: UUID) {
        items.removeAll(where: { $0.id == id })
    }
}


